Keep track of the parent alias when aliases are mapped

abderrahim opened the pull request:
This also adds tests for non-identity alias mapping and interaction between alias mapping and source mirrors.

abderrahim committed:
tests/frontend/mirror.py: Test non-identity alias mapping

abderrahim committed:
tests/frontend/mirror.py: Test interaction of alias mapping and source mirrors

abderrahim committed:
Keep track of the parent alias when aliases are mapped

This fixes the test in the previous commit

juergbi commented on src/buildstream/_project.py:
```diff
@@ -453,7 +453,7 @@ def alias_exists(self, alias, *, source, first_pass=False):
     def get_alias_uris(
         self, alias: str, *, first_pass: bool = False, tracking: bool = False
-    ) -> Sequence[Union[SourceMirror, str, None]]:
+    ) -> Sequence[Union[Tuple[str, Union[SourceMirror, str]], None]]:
```
Can we define a `NamedTuple` to make this more readable?

abderrahim committed:
Use a dataclass for the alias substitution

It is starting to get out of hand with the inclusion of the effective
alias. It is also a good idea to signal to the users that this is an opaque
type that plugin authors should not look into its internals.

gtristan commented on tests/frontend/mirror.py:
```diff
@@ -895,9 +895,29 @@ def test_source_mirror_plugin(cli, tmpdir):
 @pytest.mark.parametrize("disallow_subproject_uris", [True, False])
 @pytest.mark.parametrize("fetch_source", ["aliases", "mirrors"])
-@pytest.mark.parametrize("alias_override", [["foo"], ["foo", "bar"], "global", "invalid"])
+@pytest.mark.parametrize(
```
Can we have a comment around here explaining why this level of parameterized testing is needed, and what defects we are specifically protecting against ?

abderrahim replied:
This is mainly to be able to exclude some elements from the test matrix.

Initially I wanted something like
```
@pytest.mark.parametrize("alias_override", [["foo"], ["foo", "bar"], "global"])
@pytest.mark.parametrize("alias_mapping", ["identity", "project-prefix", "invalid"])
```

but "global,project-prefix" isn't implemented yet (I'd like to have it in a future release) and "global,invalid" doesn't make sense.

As for the reason I'm doing this, the bug in question happens when having a non-identity mapping along with source mirror plugin. When I started, I thought it could be unrelated to source plugins, so wanted to test it.

If you have a better option, feel free to comment. Or I can add the above to a comment?

juergbi replied:
Maybe calling `pytest.xfail()` for these two cases (with a comment) would be reasonable.

abderrahim replied:
Added a comment.

abderrahim committed:
tests/frontend/mirror.py: add comment to explain the alias mapping test
